**References:**

[1]	“Python Lists.” https://www.w3schools.com/python/python_lists.asp (accessed Jan. 29, 2021).

[2]	“Python Arrays.” https://www.w3schools.com/python/python_arrays.asp (accessed Jan. 29, 2021).

[3] "joeyajames/ udemy_data_structures" https://github.com/joeyajames/udemy_data_structures (accessed Feb. 1, 2021).

[4]	“Learning with Python_ How to Think Like a Computer Scientist” https://drive.google.com/file/d/1S_q1N7gZBXHzrl9p9p6EiFUlem8owRoi/view?usp=sharing 





# **BASIC EXERCISE**

**Question 84, 85, 86, 87:** Write a function that imports a one-dimensional array of real/interger numbers and prints it on the screen.

In [ ]:
n = int(input())
lst = []
for i in range(n):
    lst.append(int(input()))
print(lst)

3
1
2
3
[1, 2, 3]


In [ ]:
n = int(input())
lst = ([float(input()) for i in range(n)])
print(lst)

3
1.9
-8.2
3.4
[1.9, -8.2, 3.4]


**Question 88:** Write a function that lists even values in a one-dimensional array of numbers.

In [ ]:
def list_even_values(lst):
    for val in lst:
       if val % 2 == 0:
         print(val, end = ' ')

n = int(input())
lst = ([int(input()) for i in range(n)])
list_even_values(lst)

4
5
2
9
10
2 10 

**Question 89:** Write a function that lists positions where the value is negative in a one-dimensional array of real numbers.

In [ ]:
def list_index_of_negative_values(lst):
    for i in range(len(lst)):
        if lst[i] < 0:
            print(i, end = ' ')

n = int(input())
lst = ([float(input()) for i in range(n)])
list_index_of_negative_values(lst) 

3
-3
2
-3
0 2 

**Question 90:** Finds "maximum value" in a one-dimensional array of real numbers

In [ ]:
lst = [-5, 3, 2.8, 2, 9.5, 7, -3.5]
max(lst)

9.5

**Question 91:** Finds the "first positive value" in a one-dimensional array of real numbers. If the array has no positive value, -1 is returned.

In [ ]:
def first_positive_value(lst):
    """Return the "first positive value" in a one-dimensional array of real numbers.
    If the array has no positive value, -1 is returned."""
    for val in lst:
        if val > 0:
            return val
    return -1

help(first_positive_value)

Help on function first_positive_value in module __main__:

first_positive_value(lst)
    Return the "first positive value" in a one-dimensional array of real numbers.
    If the array has no positive value, -1 is returned.



In [ ]:
n = int(input())
lst = [float(input()) for i in range(n)]

print(lst)
first_positive_value(lst)

5
3
-12
-1.2
8
-25
[3.0, -12.0, -1.2, 8.0, -25.0]


3.0

# **SEARCH AND LIST EXERCISES**

**Question 132:** List negative numbers in a one-dimensional array of real numbers


In [ ]:
n = int(input('n = '))
lst = [float(input()) for i in range(n)]

print([number for number in lst if number < 0])

n = 5
-5
3
-9.5
2
-1
[-5.0, -9.5, -1.0]


**Question 133:** List the values in a one-dimensional array of real numbers in a given segment `[x, y]`





> 1. Use the `split ()` method to separate data *entered* on the *same line*.
2. The `map (type, obj)` method will convert the data type of all elements in `obj` to `type`. In this example, we apply map which converts type 2 input data to ***int***







In [ ]:
n = int(input('n = '))
lst = [float(input()) for i in range(n)]
x, y = map(int, input('[x, y] = ').split())
print(lst[x : y + 1])

n = 5
-3.55
12.4
8
-12.4
79.2
[x, y] = 1 3
[12.4, 8.0, -12.4]


**Question 139** List the positons where the value is the largest in a one-dimensional array of real numbers. 



> `Enumerate()` method adds a counter to an iterable and returns it in a form of enumerate object.





In [ ]:
n = int(input('n = '))
lst = [float(input()) for i in range(n)]

print([index for index, value in enumerate(lst) if value == max(lst)])

n = 4
-30.5
12.3
12.3
5.4
[1, 2]


**Question 151:** List values in an array of integers whose first digit is an odd number.

In [ ]:
def get_first_digit(n):
    digit = str(n)
    return int(digit[0])

n = int(input('n = '))
lst = [int(input()) for i in range(n)]

print([value for value in lst if get_first_digit(value) % 2 != 0 ])

n = 4
58
20
11
89
[58, 11]


# **SUMMING TECHNIQUES**

**Question 154:** Calculate the sum of elements in the array

In [ ]:
n = int(input('n = '))
lst = [int(input()) for i in range(n)]
sum(lst)

n = 3
19
1
5


25

**Question 157:** Calculate the sum of values with the tens digit is  5 in the array of integers

In [ ]:
def tens_digit_is_five(n):
    unit = n % 10
    ten = (n % 100 - unit) // 10
    return True if ten == 5 else False

def sum_all_tens_digit_is_five(lst):
    Sum = 0
    for val in lst:
        if tens_digit_is_five(val):
            Sum += val
    return Sum

In [ ]:
n = int(input('n = '))
lst = [int(input()) for i in range(n)]
sum_all_tens_digit_is_five(lst)

n = 4
15
5755
159
231


5914

**Question 158:** Calculate the sum of values greater than the value preceding it in a one-dimensional array of real numbers.

In [ ]:
n = int(input('n = '))
lst = [float(input()) for i in range(n)]

sum = 0
for i in range(1, len(lst)):
    if(lst[i] > lst[i - 1]):
        sum += lst[i]
print(sum)

n = 3
1
2
3
5.0


**Question 164:** Calculate the sum of values that have the first digit as an even number in the array of integers

In [71]:
from numpy import random

In [23]:
def first_digit_is_even(number):
    number = str(number)
    return True if int(number[0]) % 2 == 0 else False

def sum_values_first_digit_is_even(arr):
    return sum([number for number in arr if first_digit_is_even(number)])

In [30]:
n = random.randint(30)
arr = random.randint(100, size = (n))

print(arr)
sum_values_first_digit_is_even(arr)

[89 76 12 79 23 74 46 99 78 89 99 88 28 15 79 32 78]


363

**Question 166:** Calculate *the average* of positive numbers in an array of real numbers

In [83]:
n = random.randint(10)
arr = random.uniform(-99, 99, n)

print(arr)
np.mean([number for number in arr if number > 0]) # calculate average

[-67.91253  61.99319 -16.62526  -6.9941   -3.45786]


61.99319288571044

**Question 168:** Calculate *the average multiplies* positive values in a one-dimensional array of real numbers

In [91]:
n = random.randint(10)
arr = random.uniform(-99, 99, n)

print(arr)

[ 29.11958  92.19215  60.12304  35.80923 -47.79181]


5779828.901001129

In [97]:
def calculate_average_multiplies_positive_values(arr):
    arr = [number for number in arr if number > 0]
    multiplies = np.prod(arr)
    size = len(arr)
    return multiplies / size

calculate_average_multiplies_positive_values(arr)

1444957.2252502823

# **COUNTING TECHNIQUES**

**Question 171:** Count the number of even numbers in a one-dimensional array of integer numbers.

In [ ]:
def count_even_numbers(lst):
    count = 0
    for i in range(len(lst)):
        if lst[i] % 2 == 0:
            count += 1
    return count

n = int(input('n = '))
lst = [int(input()) for i in range(n)]
count_even_numbers(lst)

n = 3
1
2
3


1

**Question 171:** Count the number of positive values divided by 7 in a one-dimensional array of integer numbers

In [ ]:
def count_positive_values_divided_by_seven(lst):
    count = 0
    for i in range(len(lst)):
        if lst[i] % 7 == 0:
            count += 1
    return count

n = int(input('n = '))
lst = [int(input()) for i in range(n)]
count_positive_values_divided_by_seven(lst)

n = 4
7
12
7
49


3

# **DELETE EXERCISES**

**Question 225:** Delete the elements with index `k` in the array. 

In [ ]:
n = int(input('n = '))
lst = [float(input()) for i in range(n)]
k = int(input('k = '))

print(lst)
del(lst[k])
print(lst)

n = 2
-3.5
2
k = 1
[-3.5, 2.0]
[-3.5]


**Question 231:** Delete all "prime numbers" in the integer array

In [ ]:
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, n):
        if n % i == 0:
            return False
    return True

In [ ]:
# Approach 1 (Error)
def delete_all_prime(lst):
    i = 0
    real_len = len(lst)
    while(i < n):
        if is_prime(lst[i]):
            lst.remove(lst[i])
            i -= 1
            real_len -= 1
        i += 1

n = int(input('n = '))
lst = [int(input()) for i in range(n)]

print(lst)
delete_all_prime(lst)
print(lst)

In [ ]:
# Approach 2
n = int(input('n = '))
lst = [int(input()) for i in range(n)]

new_lst = [val for val in lst if is_prime(val) == False]
print(lst, new_lst)

n = 5
2
5
9
5
19
[2, 5, 9, 5, 19] [9]


# **Numpy (Numerical Python)**

According to [w3schools.com](https://) :

NumPy is a Python library used for working with arrays.

It also has functions for working in domain of linear algebra, fourier transform, and matrices.



https://github.com/joeyajames/Python/blob/master/Numpy/Python%20Numpy%20Intro.ipynb


https://codelearn.io/sharing/tim-hieu-thu-vien-numpy-trong-python

In [ ]:
import numpy as np

In [ ]:
import numpy as np

n = int(input('n = '))
arr = np.array([input() for i in range(n)], dtype = 'i')

print(arr.dtype)

n = 1
1
int32


In [ ]:
def hard_min_swaps(arr):
    count = 0
    for i in range(len(arr) - 1):
        min_element = min(arr[i:])
        print(min_element, arr[i])
        if arr[i] > min_element:
            print('swap, ', arr.index(min_element))
            print(arr)
            temp = arr[i]
            arr[i] = arr[arr.index(min_element)]
            arr[arr.index(min_element)] = temp
            count += 1
            print(arr)
    return count

arr = [4,3,1,2]
hard_min_swaps(arr)

1 4
swap,  2
[4, 3, 1, 2]
[4, 3, 1, 2]
1 3
swap,  2
[4, 3, 1, 2]
[4, 3, 1, 2]
1 1


2

In [ ]:
arr = [4,3,1,2]

arr[2] = 3
print(arr)

[4, 3, 3, 2]


In [68]:
import random

print("Random float number with 2 decimal places")
num2 = round(random.random(), 2)
print(num2)

num2 = round(random.uniform(33.33, 66.66), 2)
print(num2)

print("Random float number with 1 decimal places")
print(round(random.random(), 1))
print(round(random.uniform(33.33, 66.66), 1))


Random float number with 2 decimal places
0.9
33.35
Random float number with 1 decimal places
0.5
66.6
